### ISSUE TITLE
a race condition that doesn't happen under rr chaos mode

### ISSUE THREAD

**Issue by JeffBezanson** [mozilla/rr#2362]
To reproduce:
* Check out and build commit https://github.com/JuliaLang/julia/pull/32114/commits/c35d1d7f7c9d90e0a00ded1980d6fcd5d06f5433
* Run https://gist.github.com/JeffBezanson/30c46a96f1743d3a92fbdd23779806d2 with `JULIA_NUM_THREADS=2 ./julia pfib.jl`

This reliably triggers a type error that's caused by a race condition. The fix was to switch the order of two writes, in commit https://github.com/JuliaLang/julia/pull/32114/commits/f81d4d21862c99e1c0452bdf2613ab9e20da9935.

I have not been able to get the bug to trigger under rr 5.2.0, with or without chaos mode, so I thought this case might be interesting.

While I'm here let me add that rr is amazing and has been incredibly useful; thanks!

**Comment by rocallahan** [comment]
Thank you for reporting this!

The bug is that once you set `t->state` the other thread's `wait` can return and you'll use `t->result` right away, but it might not have been set yet. For rr chaos mode to catch this, we have to be lucky enough to trigger a context switch in the window between setting `t->state` and `t->result`, and the probability of that is indeed extremely low. In fact, depending on how the hardware dispatches the PMU interrupt, it might never happen.

To fix this we would need to determine that those instructions are "interesting" and increase the probability of triggering a context switch within them. One idea I've had for that would be to scan the binaries and detect atomic instructions, and use breakpoints to trigger context switches before and after atomic instructions. Unfortunately that wouldn't have helped here since before the bug was fixed, the values were written non-atomically :-(. (This is why chaos mode improvements need to be driven by real-world use cases rather than armchair theorizing!)

Another possible approach would be to use dynamic race detection to identify potentially-racing instructions and then trigger context switches around those instructions. Has anyone ever run TSAN or something like that on this code?

**Comment by Keno** [comment]
> Has anyone ever run TSAN or something like that on this code?

No, we were gonna try that next, but then we realized what the bug was :)

**Comment by JeffBezanson** [comment]
Yes, the two writes are very close so I understand it may not be realistic to be able to handle this directly. I do like the idea of context switching around atomic ops; I imagine we will eventually use them for most/all of our thread-communicating memory accesses so it would probably be useful.
